src/frontend/src/utils/condition.rs

```diff
@@ -109,6 +109,83 @@ impl Condition {
         .unwrap()
     }
 
+    /// Split the condition expressions into (N choose 2) + 1 groups: those containing two columns
+    /// from different buckets (and optionally, needing an equal condition between them), and
+    /// others.
+    ///
+    /// `input_num_cols` are the number of columns in each of the input buckets. For instance, with
+    /// bucket0: col0, col1, col2 | bucket1: col3, col4 | bucket2: col5
+    /// `input_num_cols` = [3, 2, 1]
+    ///
+    /// Returns hashmap with keys of the form (col1, col2) where col1 < col2 in terms of their col
+    /// index.
+    ///
+    /// `only_eq`: whether to only split those conditions with an eq condition predicate between two
+    /// buckets.
+    #[must_use]
+    pub fn split_by_input_col_nums(
+        self,
+        input_col_nums: &[usize],
+        only_eq: bool,
+    ) -> (HashMap<(usize, usize), Self>, Self) {
+        let mut bitmaps = Vec::with_capacity(input_col_nums.len());
+        let mut cols_seen = 0;
+        for cols in input_col_nums {
+            bitmaps.push(FixedBitSet::from_iter(cols_seen..cols_seen + cols));
+            cols_seen += cols;
+        }
+
+        let mut pairwise_conditions = HashMap::with_capacity(self.conjunctions.len());
+        let mut non_eq_join = vec![];
+
+        for expr in self.conjunctions {
+            let input_bits = expr.collect_input_refs(cols_seen);
+            let mut subset_indices = Vec::with_capacity(input_col_nums.len());
+            for (idx, bitmap) in bitmaps.iter().enumerate() {
+                if !input_bits.is_disjoint(bitmap) {
+                    subset_indices.push(idx);
+                }
+            }
+            if subset_indices.len() != 2 || (only_eq && Self::as_eq_cond(&expr).is_none()) {
+                non_eq_join.push(expr);
+            } else {
+                // The key has the canonical ordering (lower, higher)
+                let key = if subset_indices[0] < subset_indices[1] {
+                    (subset_indices[0], subset_indices[1])
+                } else {
+                    (subset_indices[1], subset_indices[0])
+                };
+                let e = pairwise_conditions
+                    .entry(key)
+                    .or_insert_with(Condition::true_cond);
+                e.conjunctions.push(expr);
+            }
+        }
+        (
+            pairwise_conditions,
+            Condition {
+                conjunctions: non_eq_join,
+            },
+        )
+    }
+
+    /// Returns the `InputRefs` of an Equality predicate if it matches
+    /// ordered by the canonical ordering (lower, higher), else returns None
+    fn as_eq_cond(expr: &ExprImpl) -> Option<(InputRef, InputRef)> {
```

Comment:
Not sure if this part of change is necessary. We already have `EqJoinPredicate`, which can be created from `Condition`.

Reply:
Hmm, this is actually used in `split_eq_keys`, which is in turn used to derive `EqJoinPredicate`. This is just used to make `split_eq_keys` and `split_by_input_col_nums` more readable